src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs

```diff
@@ -12194,6 +12193,59 @@ public void TestAnalyzerFilteringBasedOnSeverity(DiagnosticSeverity defaultSever
             }
         }
 
+        [Fact]
+        public void SourceGenerators_EmbeddedSources()
+        {
+            var dir = Temp.CreateDirectory();
+            var src = dir.CreateFile("temp.cs").WriteAllText(@"
+class C
+{
+}");
+
+            var generatedSource = "public class D { }";
+            var generator = new SimpleGenerator(generatedSource, "generatedSource.cs");
+
+            VerifyOutput(dir, src, includeCurrentAssemblyAsAnalyzerReference: false, additionalFlags: new[] { "/langversion:preview", "/debug:embedded", "/out:embed.exe" }, generators: new[] { generator }, analyzers: null);
+
+            ValidateEmbeddedSources_Portable(new Dictionary<string, string> { { Path.Combine(dir.Path, "Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.SimpleGenerator_generatedSource.cs"), generatedSource } }, dir, true);
+
+            // Clean up temp files
+            CleanupAllGeneratedFiles(src.Path);
+        }
+
+        [Theory]
+        [InlineData("partial class D {}", "file1.cs", "partial class E {}", "file2.cs")] // different files, different names
+        [InlineData("partial class D {}", "file1.cs", "partial class E {}", "file1.cs")] // different files, same names
+        [InlineData("partial class D {}", "file1.cs", "partial class D {}", "file2.cs")] // same files, different names
+        [InlineData("partial class D {}", "file1.cs", "partial class D {}", "file1.cs")] // same files, same names
+        [InlineData("partial class D {}", "file1.cs", "", "file2.cs")] // empty second file
+        [InlineData("partial class D {}", "file1.cs", "", " \\ / : * ? \" < > | ")] // illegal path name chars
```

Comment:
Do we just silently drop the illegal path name characters?

Reply:
@tmat will portable PDB's handle path names with illegal characters okay?

Reply:
@tmat can confirm, but the PDB itself seems fine. Interestingly VS chokes a bit when hitting a file with them: it steps in, and you can actually step over each line and see the output, but no file opens in the IDE itself (no crash or visible errors though)

---
In reply to: [434767181](https://github.com/dotnet/roslyn/pull/44768#discussion_r434767181) [](ancestors = 434767181)

Reply:
I think we should file a follow up item here to resolve for 16.8. Need to take into account the following:

- How does hint path impact the physical file on disk when we're in dump to disk mode
- How do we deal with generators where added files hint paths differ by only illegal file system characters
- How do illegal characters in the path impact other parts of the experience

Don't think we necessarily need to solve this for 16.7 though. One simple approach though is just ban illegal file system characters in hint paths because at the end of the day this is meant to represent a file name on disk :smile:

Reply:
Portable PDB has no requirements on path strings. Note that all characters except for `\0` are valid path characters on Linux.

Reply:
BTW, interesting test could be incomplete Unicode surrogates. PDB should handle them as well (and preserve their value). Not sure what happens elsewhere in the system though.